shelley/chain-and-ledger/executable-spec/src/Cardano/Ledger/SafeHash.hs

```diff
@@ -0,0 +1,164 @@
+import Data.ByteString (ByteString)
+import Data.ByteString.Short (ShortByteString, fromShort)
+import Data.MemoBytes (MemoBytes (..))
+import Data.Typeable
+import NoThunks.Class (NoThunks (..))
+
+-- ==========================================================
+-- A SafeHash is a hash of something that is safe to hash.
+
+newtype SafeHash crypto index = SafeHash (Hash.Hash (CC.HASH crypto) index)
+  deriving (Show, Eq, Ord, SafeToHash, NoThunks, NFData)
+
+deriving newtype instance HeapWords (Hash.Hash (CC.HASH c) i) => HeapWords (SafeHash c i)
+
+deriving instance (Typeable index, CC.Crypto c) => ToCBOR (SafeHash c index)
+
+deriving instance (Typeable index, CC.Crypto c) => FromCBOR (SafeHash c index)
+
+-- We do NOT export the constructor SafeHash, but instead export
+-- 'hashAnnotated' and 'extractHash' which have constraints that limit
+-- their application to types which preserve their original serialization
+-- bytes. These are the functions people use.
+
+type HasAlgorithm c = Hash.HashAlgorithm (CC.HASH c)
+
+hashAnnotated :: forall c i x. (HasAlgorithm c, HashAnnotated x i c) => x -> SafeHash c i
+hashAnnotated x = hashProxy (Proxy @c) (indexProxy x) x
+
+extractHash :: SafeHash crypto i -> Hash.Hash (CC.HASH crypto) i
+extractHash (SafeHash h) = h
+
+-- Occasionally we need to change the HashAnnotated (which is a phantom type)
+
+castSafeHash :: forall i j c. SafeHash c i -> SafeHash c j
+castSafeHash (SafeHash h) = SafeHash (Hash.castHash h)
+
+-- Don't use this except in Testing to make Arbitrary instances, etc.
+unsafeMakeSafeHash :: (Hash.Hash (CC.HASH crypto) index) -> SafeHash crypto index
+unsafeMakeSafeHash x = SafeHash x
+
+-- =====================================================================
+-- Only Types that preserve their serialisation bytes are members of the
+-- class SafeToHash. There are only a limited number of primitive direct
+-- instances of SafeToHash, all present in this file. We never add new
+-- instances except by deriving for a newtype (around a type S) where
+-- their is already an instance: (SafeToHash S). We can also make
+-- a (HashAnnotated T Tag) instance, which sets the phantom index. This is
+-- easy to do since its only method has a default instance. Here is a example:
+--
+-- newtype T era = T S
+--    deriving Eq
+--    deriving newtype SafeToHash  -- Uses {-# LANGUAGE DerivingStrategies #-}
+-- instance HashAnnotated (T era) Index (Crypto era)
+--
+-- So after these declarations. one specialization of 'hashAnnotated' is
+-- hashAnnotated :: Era e => T e -> SafeHash (Crypto e) Index
+-- =====================================================================
+
+class SafeToHash t where
+  originalBytes :: t -> ByteString
+  hashWithProxies :: Era e => Proxy e -> Proxy index -> t -> SafeHash (Crypto e) index
+  hashWithProxies _ _ x = SafeHash $ Hash.castHash (Hash.hashWith originalBytes x)
+  hashProxy :: Hash.HashAlgorithm (CC.HASH c) => Proxy c -> Proxy index -> t -> SafeHash c index
```

Comment:
These names are very confusing! They take exactly the same number of proxies, and hash neither of them!

I suggest:

`hashWithEraProxy`
`hashWithCryptoProxy`

Reply:
Since these function are only used in two files outside this file I went with making one function called 
makeHashWithExplicitProxys.